### ISSUE TITLE
Extract Test Framework Adapters to separate composer packages

### ISSUE THREAD

**Issue by maks-rafalko** [infection/infection#922]
Work is based on the following resources

* [Principles of OOD](http://butunclebob.com/ArticleS.UncleBob.PrinciplesOfOod), Robert Martin
* [Principles of Package Design](https://www.apress.com/us/book/9781484241189), Matthias Noback

This is the RFC with detailed description of Package Design Principles applied to Infection, that show the benefits of splitting `infection/infection` to separate packages. The main topic here is Test Framework Adapters, but splitting Infection is not limited to them (see another [RFC for extracting Mutators](https://github.com/infection/infection/issues/669)).

For me, there are no doubts we will split it. The question is how deep we will go.

# Principles about package cohesion

## Common Reuse Principle

_Classes that are used together are packaged together_

We clearly violate this priciple. People who use only `PHPSpec` adapter should not download `Codeception`/`PHPUnit` adapters.

At the moment, `infection/infection` contains classes that are not all used by everyone who uses Infection package in their project, so it violates the _Common reuse principle_. It also contains classes (the same classes actually) that are not closed against the same kinds of changes, so the package violates the _Common closure principle_. More on it later.

### Dependencies of Test Framework

Another thing that shows we violate this package principle is that `infection/infection` [requires](https://github.com/infection/infection/blob/7adc1fa1fdab6f33734249787082f094e00f16a9/composer.json#L60) `symfony/yaml`. The key thing is that Infection has nothing to do with YAML. It's a dependency of `Codeception` and `PHPSpec` adapters, but not Infection itself. Each time you use only `PHPUnit` adapter (and don't use `Codeception`/`PHPSpec`), you have

* the code you don't really need, that is not executed
* one more reason to be affected by additional dependency
* one more possible conflicting dependency when install Infection via Composer; and so on

### Cohesion

When we extract each Test Framework Adapter to a separate package, they will be coherent: all the classes a separate package contains are about the same thing. When user needs `PHPUnit` adapter, all the classes from `infection/phpunit-adapter` will be used together. And *no classes* will be used and downloaded from other adapters.

## The Common closure principle

* _The classes in a package should be closed together against the same kinds of changes. A change that affects a package affects all the classes in that package._
* _The primary justification for this principle is that we want a change to be limited to the smallest number of packages possible._

When a new version of the package becomes available, the user will upgrade their project (`composer update infection/infection`) to require the new dependency version. But they only want to do so if the changes **we** made to the package have something to do with the way the package is used in **their** project, since every upgrade requires them to verify that the code still works correctly with the new version of our package.

What does it mean to us? When we, for example, fix a bug for the code related to `Codeception` adapter, people who use only `PHPSpec` adapter should not be affected. They should not download the new version of `infection/infection` (this is how it works currently with the monolith), which may or may not have unwanted side-effects or just bring many unrelated changes from the repository. Instead, only `infection/codeception-adapter` would be updated.

> As a package maintainer you should follow the _Common closure principle_ to prevent yourself from “opening” a package for all kinds of unrelated reasons. It helps you prevent bringing out new releases that are **irrelevant** to most of your users. With this goal in mind the principle advises you to put classes in _different packages_ if they have _different reasons to change_.
>
> Page 153 - Principles of Package Design

Another positive benefits:

* we will be able to set version constraints for any dependency. See this my commit https://github.com/infection/infection/commit/73bb17a0f98d6fda67a276d29648e5a54170767d#diff-d9b88806f4deaae39c43c22fb7f31d0fR214-R228: it checks that `codeception/codeception` must be of `^3.1.1` version. This should not be done in the main repository, but only in adapter
* we will be able to track packages statistics separately:
    * how many downloads `infection/phpunit-adapter` has, the growing (or not) [installs diagram](https://packagist.org/packages/infection/infection/stats)
    * is it popular or not (stars, dependents), etc.
* we will be able to release test framework adapters *independent* of `infection/infection`. This is a big win since people who use `PHPUnit` only will not be affected by a bugfix (or new introduced bug, who knows) in the recently updated `PHPSpec` adapter's code

In this section about cohesion in Package Principles, I conclude that we need to extract the following 3 separate packages from `infection/infection`:

* `infection/phpunit-adapter`
* `infection/phpspec-adapter`
* `infection/codeception-adapter`

# Principles about packages coupling

Imaging we already extracted Test Framework Adapters to separate packages. Now, let's think what for example `infection/phpunit-adapter` will depend on. Here we have a couple of ways to go:

1. `infection/phpunit-adapter` depends on `infection/infection` and uses `TestFrameworkAdapter` interface located in the main `infection/infection` core package
2. `infection/phpunit-adapter` depends on a small, abstract and highly stable package, e.g. `infection/abstract-testframework-adapter` wich contains only abstractions and probably DTOs for Test Framework Adapters (or `infection/contracts`, read below)

The following coupling principles will help us determining which of the ways above is more suitable for Infection.

## The Stable Dependencies Principle (SDP)

Let's start from definitions.

* _Stable package_ has no or few dependencies it depend on
* _Unstable package_ needs to be updated each time its dependencies upgraded
* _Responsible package_ has many dependents - packages that depend on _this_ package
* _Irresponsible package_ is not used by any packages

> Packages that are more independent and responsible should be considered _highly stable_. Those are packages that don’t need to change because of a change in one of their dependencies, but they also can’t easily change themselves because other packages heavily depend on them.

And here is the Stable Dependencies Principle:

_The dependencies between packages in a design should be in the direction of the stability of the packages. A package should only depend upon packages that are more stable than it is._

Actually, we can calculate the "instability value" and see how our packages can be designed and how this plays with this package design principle.

### Instability value

There is a formula to measure _Instability_ of any package.

_I = C<sub>out</sub> / (C<sub>in</sub> + C<sub>out</sub>)_

where 

* _I_ - instability
* _C<sub>out</sub>_ - the number of classes _outside_ the package that any class _inside_ the package depends upon
* _C<sub>in</sub>_ - the number of classes _outside_ the package that depend on a class _inside_ the package.

_I_ value will be between `0` and `1` where `1` indicates that the package is maximally _unstable_ and `0` indicates that it is maximally _stable_.

A highly _stable_ package is _responsible_: it has many dependents, so C<sub>in</sub> is a high number. A highly _unstable_ package is very _dependent_. It has many dependencies, so C<sub>out</sub> is a high number. 

How can our packages be designed? There are at least 2 possible ways.

#### 1. `TestFrameworkInterface` is inside `infection/infection`

The first case, when the `TestFrameworkInterface` is placed inside `infection/infection`, and our Test Framework Adapters packages depend on `infection/infection`.

![test-framework-adapter(1)](https://user-images.githubusercontent.com/3725595/71320291-4b0ece80-24ba-11ea-9d6f-cc303b07568f.png)

From the first look, this is a simple and working solution. But there is a big issue: if the `infection/codeception-adapter` would depend on `infection/infection` to get the `TestFrameworkAdapter` interface from there, we would need to update its dependencies each time `infection/infection` is changed, even when the change is absolutely unrelated to the adapter's behaviour.

This is how the `infection/codeception-adapter`'s composer.json would look like:

```json
{
    "require": {
        "symfony/yaml": "^3.4.29 || ^4.0 || ^5.0",
        "infection/infection": "^0.13 | ^0.14 | ^0.15 ... and so on"
    }
}
```

Each time a new `infection/infection` version is released (even with the changes irrelevant to `infection/codeception-adapter`), we will need to update `composer.json` file to make this adapter compatible with the new version. And each such update means we need to test it to avoid potentially broken integration between adapter and the core Infection. We don't know, whether the new release of `infection/infection` affects us or not, but we still need to adopt these changes, and `composer outdated` will really show that adapter's dependency is not up to date.

We will end up with the similar PRs as [this one](https://github.com/infection/infection/pull/885), constantly updating `infection/infection`'s versions.

What about _Instability value_ for this way of package relations?

_I = C<sub>out</sub> / (C<sub>in</sub> + C<sub>out</sub>)_

![instability-first](https://user-images.githubusercontent.com/3725595/71320303-6d085100-24ba-11ea-90e3-5aa9bff74245.png)

Note: [14 is a number of packages](https://github.com/infection/infection/blob/169d9c3f5ec62b84de137fec83248ba1ae95d4da/composer.json#L49-L63) `infection/infection` is currently depends on.

This way of packages design is clearly violates the Stable Dependencies Principle, because `infection/codeception-adapter` with _I=0.5_ depends on a package `infection/infection` with a bigger _I=0.875_. But according to the Stable Dependencies Principle, the dependencies between packages should be “in the direction of the stability of the packages" (or "in the direction of decreasing Instability").  Hence a red line.

#### 2. `TestFrameworkAdapter` is inside highly stable `infection/abstract-testframework-adapter`

So, we need to make `infection/codeception-adapter` (and other adapters) to depend on something more stable, more reliable. Here is what we can do:

![abstract-test-framework-adapter(1)](https://user-images.githubusercontent.com/3725595/71320314-92955a80-24ba-11ea-8d6f-cd24cd871c45.png)

`infection/abstract-testframework-adapter` is a _highly stable_ because it depends on 0 packages. It has only 1 reason to change - changing the Public API of `TestFrameworkAdapter`.

`infection/codeception-adapter`'s `composer.json` in this case would be like this:

```json
{
    "require": {
        "symfony/yaml": "^3.4.29 || ^4.0 || ^5.0",
        "infection/abstract-testframework-adapter": "^1.0.0"
    }
}
```

Now, when the new version of `infection/infection` is released, _our adapters are not affected_. The really one reason to update adapter's dependency is when we change our abstraction, which is a rare thing.

Here is how _Instability values_ look like:

![instability-second](https://user-images.githubusercontent.com/3725595/71320317-9de88600-24ba-11ea-815c-89cfed8e769e.png)

In this example each package depends on _more stable_ packages, which is what this principle tells us to do.

There is a similar example that Mattias describes in his book about `gaufrette/filesystem` package:

> [...] After we moved the `Gaufrette\Filesystem` class and the `Gaufrette\Adapter` interface out of the main package and into a small and very stable package, it became a lot safer to use that class and/or interface in another package because the freshly created `knplabs/gaufrette-filesystem-abstraction` package is very stable.
>
> [...] The lesson to be learnt here is that when you have the ability to split a package into multiple packages with different levels of stability, you should do it. You will make it easier for people to use your code in their projects, without introducing any more instability.
>
> Page 216 - Principles of Package Design

## The Stable Abstractions Principle (SAP)

_Packages should depend in the direction of abstractness._

_Packages that are maximally stable should be maximally abstract. Instable packages should be concrete. The abstraction of a package should be in proportion to its stability._

This principle again makes sure that the more _abstract_ a package we depend on, the less chances it will be changed. Because usually changes are done in concrete classes/packages, but not abstract. Abstract packages supposed to be _stable_. Over a longer period of time it will stay the same.

As in the previous principle, we can calculate the value of _Abstractness_.

_A = C<sub>abstract</sub> / (C<sub>concrete</sub> + C<sub>abstract</sub>)_

I will show the values only for the second case, when we have a `infection/abstract-testframework-adapter`:

![abstractness-second](https://user-images.githubusercontent.com/3725595/71320319-a771ee00-24ba-11ea-9c82-c508e7b50d29.png)

As you can see, `infection/abstract-testframework-adapter` is very stable (I=0) and highly abstract (A=1). This means that this package conforms to both Stable Dependencies and Stable Abstractions principles and is a fundamental block of the package hierarchy.

By the way, it's ok adapters be non abstract (and quite instable), because such packages are real concrete implementation of some abstraction. And there is no doubt they will change during the time.

What happens if an interface `TestFrameworkAdapter` is part of a highly unstable package `infection/infection`? The answer is:

> [...] Then it is consequently not safe to depend on that package. The unstable package is likely to change. _The interface inherits the instability of its containing package_.
>
> Page 221 - Principles of Package Design

That's it.

## The Acyclic Dependencies Principle

_The dependency structure between packages must be a directed acyclic graph, that is, there must be no cycles in the dependency structure._

We don't have another packages yet so we are not affected by cycles between packages ;)

### `infection/contracts` repository

Now, we see that we need to extract Test Framework Abstraction to a separate repository. But this is not the last one. We will have at least one more abstraction - `MutatorInterface`. So it will *probably* be convenient to not have 2 separate repositories for each abstraction, but one `infection/constract`. Anyway, I'm not suggesting to create it right now, let's start with `infection/abstract-testframework-adapter` and then we will see...

### `PHAR` prefixing and bundling all adapters together

You can ask how are we goin to prepare our PHAR file after splitting Infection to different repositories and extracting Test Framework Adapters. The answer is very simple - we will always bundle Codeception, PHPUnit, PHPSpec repositories inside PHAR file. PHAR users even won't notice this splitting.

### Autodiscovering of Packages

For those who uses Infection as a composer dependency.

What is really interesting, is how we can install different "plugins" for Infection, be it Test Framework Plugin or other future plugins like custom Mutators. There is an approach of autodiscovering via composer packages, for example:

* https://psalm.dev/docs/running_psalm/plugins/authoring_plugins/#authoring-composer-based-plugins
* https://github.com/phpstan/extension-installer

The idea is very simple. A package has a special `type` inside `composer.json` file, for example `"type": "infection-plugin"`. 

```js
{
    "name": "infection/codeception-testframework-adapter",
    "type": "infection-plugin",
    "require": {
        "php": "^7.2.9",
        "ext-dom": "*",
        "ext-libxml": "*",
        "symfony/yaml": "^3.4.29 || ^4.0 || ^5.0"
    },
    "conflict": {
        "codeception/codeception": "<3.1.1" // isn't it cool?
    },
}
```

Then, Infection analyzes `composer.lock` file and autoregister all Infection plugins. The end user will just do the following change in the project's `composer.json`:

```diff
"infection/infection": "^0.15",
+ "infection/codeception-testframework-adapter": "^1.0.0"
```

and it's done - Infection automatically finds a registers adapter.

Of course, we will be able to add plugins in `infection.json` file as well. But autodiscovery is a cool feature.

### Last words

We don't need to treat our abstract packages with extension points immediately final and stable for users from the very first release. No. This is just the first step to be able to use flexible and useful Packages. And only after we quite stabilize them, we can mark them public (it's not the same as `1.0.0`) and generally available. Until then - these packages can be "private".

At this point, I hope nobody have any doubts that we have to split Infection.

**Comment by BackEndTea** [comment]
I mentioned this briefly in the slack:

I dont' like the idea of having to add multiple composer dependencies, just to get a tool working. I understand that if i want specific things, i may need to add additional dependencies. But for the 'basic' functionality, i should be able to do a single composer require, and be ready to go.

For example, if i run `composer require phpstan/phpstan`, i can then analyse my code base. I don't need to do an additonal require just to get started. Now if i want support for the assertions from `webmozart/assert`, then i need a phpstan extension/plugin. 

I do agree with you that a 'meta-package' simply containing infection and 1 adapter may be overkill, but i don't know another good way of solving this. Prehaps phpunit could be bundled with infection by default. As (for me) that is the base case of infection.

**Comment by gmponos** [comment]
Going through your description very roughly I don't disagree with it... what interests me because I am not aware of a solution and I want also to apply it to a personal proprietary project is how will you ship the `infection.phar` with the correct adapter?

**Comment by maks-rafalko** [comment]
> Prehaps phpunit could be bundled with infection by default. As (for me) that is the base case of infection.

Probably. 

Another approach, that I also mentioned in slack:

To improve DX, we can install needed packages automatically during configuration process, when Infection asks questions before it creates `infection.json.dist`

So when the user chooses PHPUnit, Infection automatically runs `composer install infection/phpunit-testframework-adapter`.

There is also a definition of "virtual package" https://devedge.wordpress.com/2014/09/27/composer-and-virtual-packages/, but I didn't have a chance to work with it and seems like there are [some disadvantages](https://matthiasnoback.nl/2014/10/composer-provide-and-dependency-inversion/)

> how will you ship the infection.phar with the correct adapter?

I've described it in the RFC - we will ship all 3 adapters in PHAR file.

**Comment by gmponos** [comment]
:/ My bad.. as I said I went very quick through it and I missed it sorry

**Comment by BackEndTea** [comment]
> To improve DX, we can install needed packages automatically during configuration process, when Infection asks questions before it creates `infection.json.dist`

I think that is the best solution, providing the best DX.

**Comment by theofidry** [comment]
Hi, I couldn't read in as much details as I wanted or as it deserves so I will definitely come back on it. Just a few thoughts:

- I am 100% behind the idea: my only points of disagreement might be on an implementation detail or two or the time/conditions we do it in
- I agree with @BackEndTea regarding the DX. You _don't_ want to install `infection/infection`, run it, and have a failure/warning that you need to install something else. This is seriously going to hurt the adoption and annoy users. You raise a good point of the awareness of requiring unnecessary code, but the truth is (and Laravel and `symfony/symfony` shows it) that there is a great deal of people that are willing to not care for the convenience it brings. That said, this is not a problem without a solution, a few that I can think of:
  - When running `Infection`, it can do "discovery" step to detect a specific test framework and then give the user the command to run if he desires to install it via a prompt or something... A bit like we do right now to create the configuration file on the first run
  - Create an installer and recommend the installer which can do that prompt/discovery step above
  - Keep `infection/infection` with all the adapters and have `infection/engine` besides the test framework adapter packages. And for the picky they can go with the engine package and the test framework of their choices and for the lazy people that just want something working without a care about unnecessary code just keep `infection/infection`. A good example of this in the wild is Symfony
  - Go with it and recommend to use the PHAR anyway...
- I am happy with a split, as long as we have the necessary tooling to make it work seamlessly. Splitting `infection/infection` in 3 or 4 packages without a sub-split a la Symfony or Laravel is a straight up PITA: it will hurt drastically contributions from both current maintainers & new maintainers. On a side note, this is also a reason why I'm really not keen to take a look at the other Infection packages (e.g. the bot or the doc) and why I'm feeling anxious whenever I have a sync change to do between php-scoper & box (although it got a lot better lately since it reach a good level of stability - it is still a recurrent point of pain though) and even more so with alice related packages

Regarding the last point as I'm sounding quite negative on it: I do want to put the emphasis on that there _is_ (multiple) solutions to this. It however definitely require some time to look it up and try it out. In other words while possible it is not a trivial task either.

Least but last I would love if we go for any split if that could wait for me go skim through the codebase. I think I've gone through 1/3 already so it's probably a matter of months which I don't think is too much to ask to try out and sort out the details of this RFC.

**Comment by maks-rafalko** [comment]
> You don't want to install infection/infection, run it, and have a failure/warning that you need to install something else.

I briefly mentioned it in [this comment](https://github.com/infection/infection/issues/922#issuecomment-568258565). Let me underline it again: we **will not** downgrade to this inconvenience. Infection will be executed by developer only 1 time (same as it is right now). But under the hood it will (1 of the ideas) run `composer require infection/xxx-testframework-adater` **automatically**, so people will not be annoyed.

This is definitely something I will dig in and try different ways, including

* autodetect during `ConfigureCommand` execution and install by running `composer require xxx`
* autodetect somehow subscribing on [Composer hooks](https://getcomposer.org/doc/articles/scripts.md#event-names) (I don't know wether it's possible or not yet)
* thinking about something similar as `symfony/phpunit-bridge`, when you run `bin/infection` - it can download needed packages in runtime

And I'm pretty sure these are not all the possible ways to achieve this goal.

---

> You raise a good point of the awareness of requiring unnecessary code, but the truth is (and Laravel and `symfony/symfony` shows it) that there is a great deal of people that are willing to not care for the convenience it brings.

Let me disagree here. 

1. I should say that there are many packages, including Symfony ones that are bad from Package Design point of view, and "Principles of Package Design" book is full of such examples. This is from historical reasons, no blame to authors of course.
2. Regarding `symfony/symfony` - did you guys know that this is deprecated and won't be tagged anymore? Please see this issue https://github.com/symfony/symfony/issues/31484#issuecomment-491684412 and in particular [this comment](https://github.com/symfony/symfony/issues/31484#issuecomment-491684412) from Fabien:

> [..] You will need to require individual packages in your `composer.json` file instead of requiring `symfony/symfony` [..]

So, `symfony/symfony` is a legacy meta package that is not allowed in Symfony >= 5 anymore and must be replaced with concrete packages.

---

Regarding monorepo vs multirepo: at this moment, I'm not so excited about monorepo, and I'm quite happy with multiple repos here in Infection. 

Anyway, be it monorepo or not, this is not about Package Design, and should be developed / discussed / investigated / analyzed in a separate ticket. I've just spent ~1h reading about monorepo and it has [as many downside as multiple repos](https://medium.com/@mattklein123/monorepos-please-dont-e9a279be011b) as well - but again, let's move this discussion out of this RFC to avoid the mess.

---

> Least but last I would love if we go for any split if that could wait for me go skim through the codebase. I think I've gone through 1/3 already so it's probably a matter of months which I don't think is too much to ask to try out and sort out the details of this RFC.

Unfortunately, I can't afford it. Let me explain:

* [3 weeks ago](https://github.com/infection/infection/pull/650#issuecomment-561557885) I started investigating package design, pushing back all my important activities to get it solved once and for all ;) And now, when I have all this information fresh in my head, when I have time and energy - I will proceed with splitting it until we get at least Test Frameworks extracted. I can't delay it anymore as switching context after weeks of pauses eats so much our (mine) time
* I'm going to prepare a talk about package design and this is what it will be based on - on a real world example of splitting Infection

Your current activities in no way conflicts with what is described in this RFC, these activities are non-blocking. I'm pretty sure we *will* have updates in any of the extracted interfaces, but that's ok. Extracted packages will be private during this experimenting.

TL;DR 

* I will continue working on it
* I will start with different experimenting repos (just for simplicity). This doesn't block us from moving to monorepo later, especially when there are a set of tools that can merge the history and do whatever we want

**Comment by sanmai** [comment]
The problem here as I see it is not because Infection has things that need other things, but that Infection depends on these other things _by default_. E.g. you may never use anything other than PHPUnit, but Infection still depends on `symfony/yaml`.

One solution is to split Infection into sub-packages, as in actually splitting the code. This is problematic due to circular dependencies appearing during testing as I outlined in
https://github.com/infection/infection/issues/930#issuecomment-569197905

I propose a different solution. Keep the code exactly same, but instead split the dependencies. E.g. for a package named `infection/some-testframework-adapter` there will be a repository with a single `composer.json` with no code whatsoever, so that Infection itself won't depend on things it does not require for itself, but when a user needs to work with certain framework, all they would need is to install that meta-package pulling missing dependencies.

Conversely, Infection will have a dev-dependency on all of these adapter packages, and therefore every change will be tested against all possible odds, as it is now.

**Comment by maks-rafalko** [comment]
> The problem here as I see it is not because Infection has things that need other things, but that Infection depends on these other things _by default_. E.g. you may never use anything other than PHPUnit, but Infection still depends on `symfony/yaml`.

Fortunately, this is not true. If you search for `Yaml`, you will notice that the only 1 thing that depends on `symfony/yaml` is a PHPSpec adapter. _This is not_ a dependency of `infection/infection`

Extreme example: imagine Codeception adapter requires `ext-mongodb` extension. Does it mean that all `infection/infection` users must have `ext-mongodb` installed? It would be a nightmare. The only users who really need it are those who in fact use Codeception adapter, but not the others.

And I would like to underline it again - there is no such thing as "default" test framework. Literally, Infection is a framework that can run project's suite via adapters. And the fact that those adapters are yet in the main package does not make them "default".

> circular dependencies

there is no circular dependencies in the design shown in the RFC. Opposite, all dependencies are inverted.

> I propose a different solution. Keep the code exactly same, but instead split the dependencies. E.g. for a package named infection/some-testframework-adapter there will be a repository with a single composer.json with no code whatsoever, so that Infection itself won't depend on things it does not require for itself, but when a user needs to work with certain framework, all they would need is to install that meta-package pulling missing dependencies.

Several issues here (again, they are covered in the resources I mentioned above)

1. In this case the code would not pass the "linking process" (imaging PHP has one as in other compiling languages). It means, that Infection will contain the code that can't actually work, because dependencies are not installed (example: PHPSpec inside core, but dependencies are not installed)
2. This solution still forces our users to pull new irrelevant updates for them (PHPUnit users will pull new updates that were made for PHPSpec adapter)
3. I hardly imagine how to support the code where its dependencies are in the other package. This sounds wrong for me.

**Comment by sanmai** [comment]
1. There isn't a linking process, but we don't need it because when I'm using Infection with PHPUnit and only, all code that touches PHPSpec-related things is a _dead_ code. It will never be executed, hence if I don't install `symfony/yaml`, I would not even notice.
2. That is a fair sacrifice to me. What it's like, 20 kilobytes, 100 kilobytes?.. Nothing is more costly than the time of our fellow developers, and these kilobytes cost nothing these days.
3. Just add these meta-packages as dev-dependencies, and there you go: everything works just as it is now, before that change. OK, you'll need to update these sub-packages from time to time, but isn't that a part of what you're proposing anyway?..

There's one drawback I had to mention: we'll have to have checks here and there for missing libraries. E.g. in code using `Yaml` we'll have a throw an exception if it isn't installed. And that will be basically untested code. Again, I think this is a fair compromised for not having people pull `symfony/yaml` etc. when they don't need them.

**Comment by maks-rafalko** [comment]
> That is a fair sacrifice to me. What it's like, 20 kilobytes, 100 kilobytes?

I'm not speaking about the size, I'm speaking about the need of pulling changes that are irrelevant to you (so you have to upgrade the dependency without a reason for it), about dependencies that are irrelevant to you (example with `ext-mongodb` clearly shows the issue).

Well, I notice that I constantly repeating myself. All of this I described with the links to files and commits in the RFC, I have nothing to add here.

Each conclusion from the RFC is supported by an example of concrete issue. To not loose the track, I suggest to discuss particular sentence, paragraph, point, statement from this RFC and say where it's wrong.

**Comment by sanmai** [comment]
> so you have to upgrade the dependency without a reason for it

That's my point! If PHPSpec adapter depends on `ext-mongodb`, then just don't install that adapter, and Composer wouldn't bug you about missing `ext-mongodb`. You don't need to split actual code into sub-projects to archive just that!

Notice that I'm totally sidetracking the discussion because my proposal _does not require_ to split the code, at all. I will try to rephrase myself as many times as needed. I can even call you or anyone on the phone, if this goes down to that.

**Comment by maks-rafalko** [comment]
TL;DR

I did one more round of reading, investigating, experimenting and analyzing everything related to this subject, and here is what I came up with:

As a first step, we will extract only Codeception and PHPSpec adapters out of Infection. PHPUnit will stay inside the core

* This will be not so agressive change, and for most of the Infection users, as well as contributors, this will be a very smooth change
* If you look at the most volatile files (`git log --pretty=format: --name-only | sort | uniq -c | sort -rg | head -30`) you can notice that we still update PHPUnit's related stuff quite a lot. Leaving PHPUnit adapter inside the core (for now) will give us some time to think about what we can do else/better
* Codeception and PHPSpec will definitely be extracted, and this extraction will solve 90% of my concerns from this RFC (we will have clear and real dependencies, pluggable adapters, etc.)


---

Answers to @sanmai [questions](https://github.com/infection/infection/issues/930#issuecomment-569221847):

First of all, let's see what is the circular dependency. When the package `A` depends on `B`, and `B` depends on `A` - it's a circular dependency (or more complex `A -> B -> C`, and `C -> A`).

On the package relations I provided in this RFC - _there are no circular dependencies_ at all. 

![tf-adapter](https://user-images.githubusercontent.com/3725595/71645063-4da3cc00-2ce3-11ea-8dec-9539af8f31de.png)

> And no integration testing against other packages, am I right?

Each adapter (Codeception, PHPSpec) will have their own e2e tests that will require `infection/infection` and test how it works with any adapter's version).

`infection/infection` in its turn will have (for now) e2e tests with PHPUnit adapter.

> So are you going to introduce `dev-master` dependencies to composer.json?

No.

Let's see how we will work with the changes in these packages. Imagine all packages are on v1.0.0 version.

|| `infection/codeception-adapter` |  `infection/infection` | `infection/abstract-testframework-adapter` |
|:-:|:-:|:-:|:-:|
|1| new change |   |   |
|2|   | new change  |   |
|3|   |   | new change  |


1. When someone updates adapter's code, in order to test these new changes - the build fetches last adapter's commit and stable versions of `infection/infection` and `infection/abstract-testframework-adapter` packages inside e2e test. So adapter's e2e tests will make sure that the new adapter's code still works correctly
2. When someone updates `infection/infection` code, only "default" adapter is tested inside e2e tests. Other adapters are not executed at this step. This is basically how it works in similar projects, for example in Flysystem. All adapters has their own testsuite with e2e tests that requires core package. But core package knows nothing about other adapters - it works with an abstraction.
3. When someone updates abstraction and releases its new version (v2.0.0), we should then update `infection/infection` to make it working with new adapter's abstraction, release it, and then update all other adapters, releasing one by one. No `dev-master` stability is needed here. It's a step by step job. No cycles as well. Each package can be updated, tested, released without a "deadlock" because of the cycles.

**Comment by theofidry** [comment]
>Each adapter (Codeception, PHPSpec) will have their own e2e tests that will require infection/infection and test how it works with any adapter's version).

Do note that this will likely require to set the `COMPOSER_ROOT_VERSION` to work out. For example when I update PHP-Scoper, I need to update it (export it or set it when using Composer) because Box depends on PHP-Scoper (PHP-Scoper only depends on it as a dev-dependency to build up the PHAR)

**Comment by maks-rafalko** [comment]
as far as I understand your case, this is because you have a cycle dependencies, and this is exactly what Composer documentation tells about `COMPOSER_ROOT_VERSION`

* [PHP-Scoper depends on Box](https://github.com/humbug/php-scoper/blob/829152ccd0a4a3cd3d02cd00bf404daf7413c718/composer.json#L33)
* [Box depends on PHP-Scoper](https://github.com/humbug/box/blob/120b0a3d45c1de5b4836fa956e2f4100887236be/composer.json#L51)

This is not the case with adapters. Adapter depends *only* on `infection/abstract-testframework-adapter`, but not on `infection/infection` (see image above).

**Comment by theofidry** [comment]
Ah indeed, I misread the dependency :)

**Comment by sanmai** [comment]
> But core package knows nothing about other adapters - it works with an abstraction.

Otherwise put, if there's an unknown breaking change in the core package, we would not know about it until someone reports it on the adapter? Or starts a build on an adapter.

Are we going to list explicit compatible versions of the core package for each adapter to work around that issue? E.g. each adapter's `composer.json` will say that it is supposed to work with versions A...B, but not higher-lower, and with each update of the core package we will have to update the adapter to list never versions. I'm not saying this is impossible to do, but it might lead to a situation where nobody's going to volunteer. 

And what if we want to improve adapters so much as they'll require changes to the core package breaking other adapter?.. This is also not impossible, but someone will have to do that, and that's not just a single PR (like it is now), but a whole bunch of them, and someone also will have to review this change. 

As the story goes, I see these issues happening [with Hoa](https://github.com/hoaproject/Central#ci-status) with the release of PHP 7.3/7.4, where there's nobody to move the ball forward: yes there are people to fix issues here and there, yet overall the project has stalled. 

(The original author of Hoa found a job somewhere at FANG, but he could have ended up under a bus just as well, whichever reasons should not be a halting factor for an open-source project.)

**Comment by maks-rafalko** [comment]
> Otherwise put, if there's an unknown breaking change in the core package, we would not know about it until someone reports it on the adapter? Or starts a build on an adapter.

We are leaving PHPUnit adapter in the core. It's just the same adapter as *any other adapter*. And if the new commit works with abstraction (which is tested with concrete PHPUnit implementation) - it should work with any other implementation because all of them must correctly implement abstraction's methods. This is related to LSP principle, and we aren't violate it in any of the adapters.

There is no need to check all the implementations for each commit to the core. We check that one "default" works and assume that other implementations work as well because they implement the same interface. And "assume" here is not something bad - this is how abstraction should work. Flysystem does not check all its adapters on each commit. It's a responsibility of adapter to correctly implement `Adapter` interface.

> Are we going to list explicit compatible versions of the core package for each adapter to work around that issue? E.g. each adapter's composer.json will say that it is supposed to work with versions A...B, but not higher-lower, and with each update of the core package we will have to update the adapter to list newer versions. I'm not saying this is impossible to do, but it might lead to a situation where nobody's going to volunteer.

Let me emphasize it again (it should be underlined by a red solid line):

* Adapter **does not** depend on `infection/infection`
* `infection/infection` **does not** depend on adapter

See the image from my [previous](https://github.com/infection/infection/issues/922#issuecomment-570076331) comment.

There is no cycle, no instable dependencies. They are only use the same abstract interface. And we shouldn't add compatible versions of adapters to the core and vice versa. I explained it in the RFC.

As long as `infection/infection` and adapter use the same interface (e.g. `infection/abstract-testframework-adapter:1.0.0`) - they work with any versions of each other. Because the only breaking change between core and adapter that can happen - sits inside the interface. Other cases will be covered by tests for PHPUnit adapter inside core, and for e2e tests inside each adapter.

> And what if we want to improve adapters so much as they'll require changes to the core package breaking other adapter? This is also not impossible

No, this is impossible. Adapter in no way influences to the core. Only abstract interface can influence to both core and adapters, and I explained this case in my previous comment.

> As the story goes, I see these issues happening with Hoa with the release

I didn't use this project, but I don't want to compare Infection with the not successful examples of the projects. Let's compare it with the Flysystem - where `AdapterInterface` has not been changed [since 2015](https://github.com/thephpleague/flysystem/blob/master/src/AdapterInterface.php) (5 years ago!).

I am 100% sure that the Package Design is not just a theoretical bullshit that is only being teached in the university by theoreticals who have never built any real packages, but instead these are the real rules to build independent, decoupled, well thought components. And I'm going to either proof it or say that this is not true to the rest of the (php) world ;).

Without trying it we can't gurantee anything. Yes, there can be limitations and difficulties, and of course from development point of view it's easier to just have 1 monolith repository and have everything coupled inside, but this is not how the sofrware should be built. And I explained the issues in RFC's description.

**Comment by theofidry** [comment]
>>As the story goes, I see these issues happening with Hoa with the release

>I didn't use this project, but I don't want to compare Infection with the not successful examples of the projects. Let's compare it with the Flysystem - where AdapterInterface has not been changed since 2015 (5 years ago!).

The problem with Hoa is the same as I personally have with alice or that I see with Zend or amp. It's not so much about the theory being wrong, it's about having a lot of dependencies to manage and not having the necessary tooling to easily do cross-packages changes which becomes an increasing burden as you get more packages or "big changes" (e.g. major deps bump, new PHP version support). Although I would say Hoa is a bit more extreme in this case.

<hr/>

As already stated: for me the proposal is good, both on the theory and how it's planned. My only reluctance is moving out code in a separate repository when I see short-term changes happening there (problem that would be solved if we were to go mono-repo).

I would really wish to hear the rest of the @infection/core team though...

**Comment by sanmai** [comment]
> My only reluctance is moving out code in a separate repository

I'd rather have the code not moved too. At least this code. 

I don't think the Package Design theory is wrong or anything, it's the matter of practically managing things to go the way the theory asks for. Splitting the code Symfony way will work, although splitting only dependencies would work just as fine, tbh.

**Comment by maks-rafalko** [comment]
> > My only reluctance is moving out code in a separate repository
> 
> I second you on this. I'd rather have the code not moved too. At least this code.

You are speaking about different things. @theofidry is ok with extracting, but in mono-repo; @sanmai is about not extracting at all.

Let me stop this discussion. I answered to all of the concerns raised above.

With all the respect, nobody said anything that breaks the theory and practice of this RFC.

If someone have something _new_ to add and check - please do it.

**Comment by sanmai** [comment]
I do think we were speaking the same. I'm OK with mono-repo approach, if it this wasn't clear.

**Comment by maks-rafalko** [comment]
It wasn't :( 

Then let's make other things clear for anyone who reads it: mono-repo still means the code is split by packages. 

Different `composer.json` files, different packages in Packagist, different namespaces, different tests, etc.

So, from the Package Design point of view - be it mono-repo or not - does not matter. Packages are split.

**Comment by sanmai** [comment]
As I understand how the mono-repo works for Symfony, is that they have one namespace, same repository, same directory for tests, and everything. But later they check out major parts of the main project into sub-projects so that you can install [`symfony/process`](https://github.com/symfony/process) without the need to pull [the entire Symfony code tree](https://github.com/symfony/symfony). But if you want to contribute, you make PRs against the main project repository, and there everything is tested against all it should be tested against. Otherwise put, you can always be sure that you're putting out completely solid code as it can be, and also you're not sacrificing almost nothing else extra. E.g. if all major Symfony developers happed to die in a plane crash (not unseen unfortunately), one could easily fork the main repository and continue as usual.

The same approach can be applied to Infection, and let @theofidry correct me if I'm thinking something wrong here. While it results in everything you expect from the Package Design -applied project, it lacks the most dangerous friction points which many other projects are plagued by.

I understand you really want to split out Infection into multiple actually separate projects, and that's certainly doable, but there can be a bitter price to pay. We see these dead or stalling projects here and there that could not afford to pay the price, and even if Infection might not be them, one has to be wary and very cautious. 

Sure there are counter-examples from the Java world, but as I saw these projects I also saw that they're being supported by paid developers. No guess their employers can afford to pay the price, else they won't be going the way too. If I were happen to be paid to hack on Infection, or any project for that matter, I'd happily consider the package design approach, for one because it's a win-win situation for a commercially maintained project as both the firm gets components for future reuse, and workers became multi-project experts that are hard to come by, hence have to be well paid and respected.

**Comment by maks-rafalko** [comment]
> As I understand how the mono-repo works for Symfony, is that they have one namespace, same repository, same directory for tests, and everything.

No, this is not how Symfony mono-repo works.

* `symfony/process` has
  * its own [`composer.json`](https://github.com/symfony/process/blob/d471271c2f2877dbed165c7130b71ef141909ab7/composer.json)
  * its own [tests folder](https://github.com/symfony/process/tree/master/Tests) and separate job to [execute tests on Travis](https://travis-ci.org/symfony/symfony/jobs/632941598#L5188-L5212)
* `symfony/filesystem` 
  * its own [`composer.json`](https://github.com/symfony/symfony/blob/20bf17f6ad38011f90cfe9b2dc7c30c8901163e5/src/Symfony/Component/Filesystem/composer.json)
 * its own [tests folder](https://github.com/symfony/symfony/tree/20bf17f6ad38011f90cfe9b2dc7c30c8901163e5/src/Symfony/Component/Filesystem/Tests) and separate job to [execute tests on Travis](https://travis-ci.org/symfony/symfony/jobs/632941598#L3992-L4017)
* and so on...

And as I [said earlier](https://github.com/infection/infection/issues/922#issuecomment-568765511), using `symfony/symfony` [is deprecated](https://github.com/symfony/symfony/issues/31484#issuecomment-491684412), so now the only right way of using components is to depend on the particular component, not on `symfony/symfony`.

I'm not sure again we are on the same page. How mono-repo works for Symfony, it can work in the exact same way for Infection. Inside monorepo there will be *different* composer packages. But again - mono-repo is not a discussion of this RFC.

> I understand you really want to split out Infection into multiple actually separate projects

I want to split it into multiple composer packages, which is stated in the title of this issue. Be it different repositories or one mono-repo with different composer packages and subtree split - doesn't matter.

**Comment by theofidry** [comment]
@maks-rafalko I think you did a good job as laying out what the RFC is about, the aim and how it solves it. And I think although there's been some communication issues or questions on the way, I think @sanmai and I are on the same page here: as long as have mono-repo, we are good with the RFC itself about splitting code and having different packages is fine.

The core of the discussion is at that point if not from almost the start actually, about #930. I'm personally very reluctant (although as I already expressed: if the rest of the team feels differently I'll happily comprise) is the polyrepo, i.e. physically splitting the code across multiple repositories.

**Comment by maks-rafalko** [comment]
Update: I've extracted Codeception Adapter out of Infection.

Currently, packages are under my personal account except one.

* `feature/extract-test-frameworks` https://github.com/infection/infection/pull/933 Infection core's changes (very simple diff)
* [infection/include-interceptor](https://github.com/maks-rafalko/infection-include-interceptor) Include Interceptor 
* [infection/codeception-adapter](https://github.com/maks-rafalko/infection-codeception-adapter) Codeception Adapter
* [infection/abstract-testframework-adapter](https://github.com/infection/abstract-testframework-adapter) Abstract Testframework Adapter

Core package still depends on `infection/codeception-adapter`. And it still depends on `symfony/yaml` because of PHPSpec, but 1 thing in one step.

Next step - remove codeception dependency form the core and implement autodiscovery of packages (aka autoregistering).

**Comment by maks-rafalko** [comment]
This ticket is no longer needed. Splitting [has begun](https://github.com/infection/infection/pull/933).
